LoRa: Initial integration of LoRa Basics Modem for LoRa API & SX126x and SX127x

JordanYates opened the pull request:
This PR contains the initial integration of LoRa Basics Modem (Semtech's replacement for LoRaMAC-node) with Zephyr.

To limit the scope, only the LoRa API is considered, and only the standalone SX126x and SX127x chipsets are ported.
The new drivers use the same devicetree definitions as the current drivers, so the only change required to use the updated drivers is `CONFIG_LORA_MODULE_BACKEND_LORA_BASICS_MODEM=y`.

Once this PR is merged, we can start to look at:
 * Basic LoRaWAN support
 * Other LoRa transceivers (stm32wl)

PR has been tested with the lora send and receive samples between a `semtech_sx1262mb2das` and `semtech_sx1272mb2das` shield in both directions.
The SX1262 transmitter, SX1272 receiver setup currently receives packets with ~10dB lower RSSI than the reverse, which I will continue to look into.

Issue for the creation of the module: #89251 (DONE).

JordanYates committed:
lora: shift drivers to dedicated `loramac-node` folder

Move the current implementation of the LoRa API using `loramac-node` to
a dedicated folder in preparation for the LoRa basics modem
implementation.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
manifest: add lora-basics-modem module

Import the lora-basics-modem module as an alternate backend for LoRa
and LoRaWAN, since loramac-node has been deprecated.

Support is currently limited and experimental.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
lora: LoRa Basics Modem backend for SX12xx chips

As a first integration of the LoRa Basics Modem backend, implement the
LoRa API for the standard SX126x/SX127x chips.

Much of the logic from `lbm_common.c` is taken from the loramac-node
`sx12xx_common` implementation, but it should now be agnostic for all
LoRa RF transceivers.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
samples: lora: validate LBM build

Add testcases for `CONFIG_LORA_MODULE_BACKEND_LORA_BASICS_MODEM` in the
LoRa API samples.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
doc: lora_lorawan: document LoRa Basics Modem

Document the current status of LoRaMAC-node and LoRa Basics Modem.

Signed-off-by: Jordan Yates <jordan@embeint.com>

kartben commented on drivers/lora/lora_basics_modem/lbm_common.h:
```diff
@@ -0,0 +1,107 @@
+ *
+ * @param dev Modem to initialise
+ *
+ * @param 0 On success
+ * @param -errno On failure
+ */
```
should actually be `@retval`, even :)

fmoessbauer commented:
I'm happy to give this a try with a non trivial application on the TTGO T-Beam board. Many thanks for doing the port to the new driver.

fmoessbauer commented:
> I'm happy to give this a try with a non trivial application on the TTGO T-Beam board. Many thanks for doing the port to the new driver.

I can confirm that this properly works on this board. Had to switch the syncword in `lbm_lora_config->params.sync_word` though, but that is specific to my use-case (which BTW is already better than before where I needed to change this in the HAL).

JordanYates commented:
> I can confirm that this properly works on this board (RX & TX). Had to switch the syncword in lbm_lora_config->params.sync_word though, but that is specific to my use-case (which BTW is already better than before where I needed to change this in the HAL).

Thanks for validating. I also have a use-case for changing the sync word, but I don't want to pile too many changes in this PR.
The only challenge will be how to add the functionality without breaking compatibility with the loramac-node backend, which will need to stick around for a while.

fmoessbauer commented:
> I don't want to pile too many changes

Sure, agree!

> The only challenge will be how to add the functionality without breaking compatibility with the loramac-node backend, which will need to stick around for a while.

I would implement that using a new API function `set_sync_word`, which just returns a `-ENOSUP` if not implemented by a driver. But we probably better move that discussion to an issue.

mirzasadiqassb commented:
Hi @JordanYates,

I'm testing this PR (#89241) on a custom board with an SX1262 for basic LoRa with the send and receive samples. I'm seeing a fatal exception and "LoRa send failed, -22" (EINVAL). This does not happen with loramac-node driver.

**My Setup:**
* **Board:** ESP32S3 with SX1262.
* **LoRa Usage:** LoRa Send/Receive samples.
* **Zephyr Version:** 4.1.99

**Relevant `prj.conf`:**
```
CONFIG_LOG=y
CONFIG_LORA=y
CONFIG_PRINTK=y
CONFIG_LORA_LOG_LEVEL_DBG=y
CONFIG_LORA_MODULE_BACKEND_LORA_BASICS_MODEM=y
```

**Console Logs:**
```[00:00:00.187,000] <dbg> lbm_driver: sx126x_hal_reset: 
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=9
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=9
--- 8 messages dropped ---
[00:00:00.219,000] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
--- 2 messages dropped ---
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0xa0 DATA_LEN=0
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8a DATA_LEN=0
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x86 DATA_LEN=0
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
[00:00:00.220,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x95 DATA_LEN=0
[00:00:00.220,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8e DATA_LEN=0
[00:00:00.220,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8b DATA_LEN=0
[00:00:00.220,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
[00:00:00.220,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
[00:00:00.220,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0
[00:00:00.220,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
[00:00:00.220,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
[00:00:00.221,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=2
[00:00:00.221,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=2
[00:00:00.221,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
[00:00:00.221,000] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP
[00:00:00.222,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0
[00:00:00.222,000] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
[00:00:05.223,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
[00:00:10.224,000] <err> lora_send: LoRa send failed, -22
```

**Request:**
My application code is the same as the samples with the addition of the return code when it failed. It would be helpful to me if you could enlighten or help me regarding this? Thanks for your work on this!

JordanYates commented:
Thanks for taking the time to try it out @mirzasadiqassb.

Based on the provided logs, it looks like the TX done interrupt is not being seen by the micro. This is interesting since the logic works on my SX1262 shield. Can you run the test again with the `LOG_DBG` in `sx126x_hal_write` and `sx126x_hal_read` commented out so we don't drop the other messages, and also provide your version of the SX1262 devicetree node?

sophiekovalevsky commented:
@JordanYates  i will definitely give it a try on this but not during this week, for sure on next week.

mirzasadiqassb commented:
> Can you run the test again with the `LOG_DBG` in `sx126x_hal_write` and `sx126x_hal_read` commented out so we don't drop the other messages, 

Here is the log when I commented out the `LOG_DBG`:
```
[00:00:00.187,000] <dbg> lbm_driver: sx126x_hal_reset: 
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP
*** Booting Zephyr OS build v4.1.0-2655-g2fd0ad6b8b6b ***
[00:00:00.218,000] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
[00:00:00.222,000] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP
[00:00:00.223,000] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
[00:00:10.224,000] <err> lora_send: LoRa send failed, -22
```


> and also provide your version of the SX1262 devicetree node?

And here is my SX1262 devicetree node:
```
lora0: lora@0 {
	compatible = "semtech,sx1262";
	reg = <0>;
	reset-gpios = <&gpio0 12 (GPIO_OPEN_DRAIN | GPIO_ACTIVE_LOW)>;
	busy-gpios = <&gpio0 13 GPIO_ACTIVE_HIGH>;
	dio1-gpios = <&gpio0 14 GPIO_ACTIVE_HIGH>;
	dio2-tx-enable;
	dio3-tcxo-voltage = <SX126X_DIO3_TCXO_1V8>;
	tcxo-power-startup-delay-ms = <5>;
	spi-max-frequency = <4000000>;
};
```

sophiekovalevsky commented:
Running some initials tests is indicating that everything is so far so good.

```
*** Booting Zephyr OS build v4.1.0-3301-g4d204c5de3c0 ***                                                                          
[00:00:00.296,295] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x9f DATA_LEN=0                                                      
[00:00:00.296,325] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE                                                   
[00:00:00.297,576] <dbg> lbm[00:00:00.593,566] <dbg> lbm_driver: sx126x_dio1_callback:                                             
[00:00:00.593,627] <dbg> lbm_driver: op_done_work_handler: op_done_work_handler: 1                                                 
[00:00:00.593,627] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x12 DATA_LEN=2                                                       
[00:00:00.593,780] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x02 DATA_LEN=0                                                      
[00:00:00.593,872] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0                                                      
[00:00:00.593,963] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP                                                             
[00:00:00.595,092] <inf> lora_send: Data sent 0!                                                                                   
[00:00:01.595,153] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0                                                      
[00:00:01.595,184] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE                                                   
[00:00:01.596,466] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1                                                       
[00:00:01.596,618] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1                                                      
[00:00:01.596,740] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8f DATA_LEN=0                                                      
[00:00:01.596,832] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0e DATA_LEN=12                                                     
[00:00:01.597,045] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x83 DATA_LEN=0                                                      
[00:00:01.597,167] <dbg> lbm_driver: lbm_lora_send: Expected airtime: 289 ms                                                       
[00:00:01.886,138] <dbg> lbm_driver: sx126x_dio1_callback:                                                                         
[00:00:01.886,199] <dbg> lbm_driver: op_done_work_handler: op_done_work_handler: 1                                                 
[00:00:01.886,199] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x12 DATA_LEN=2                                                       
[00:00:01.886,352] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x02 DATA_LEN=0                                                      
[00:00:01.886,444] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0                                                      
[00:00:01.886,535] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP                                                             
[00:00:01.887,664] <inf> lora_send: Data sent 1!                                                                                   
[00:00:02.887,756] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0                                                      
[00:00:02.887,786] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE                                                   
[00:00:02.889,068] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1                                                       
[00:00:02.889,190] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1                                                      
[00:00:02.889,312] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8f DATA_LEN=0                                                      
[00:00:02.889,434] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0e DATA_LEN=12
[00:00:02.889,648] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x83 DATA_LEN=0
[00:00:02.889,739] <dbg> lbm_driver: lbm_lora_send: Expected airtime: 289 ms
[00:00:03.178,741] <dbg> lbm_driver: sx126x_dio1_callback: 
[00:00:03.178,771] <dbg> lbm_driver: op_done_work_handler: op_done_work_handler: 1
[00:00:03.178,802] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x12 DATA_LEN=2
[00:00:03.178,924] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x02 DATA_LEN=0
[00:00:03.179,046] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
[00:00:03.179,138] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP
[00:00:03.180,236] <inf> lora_send: Data sent 2!
[00:00:04.180,297] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0
[00:00:04.180,328] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
[00:00:04.181,610] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
[00:00:04.181,732] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
[00:00:04.181,854] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8f DATA_LEN=0
[00:00:04.181,976] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0e DATA_LEN=12
[00:00:04.182,189] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x83 DATA_LEN=0
[00:00:04.182,281] <dbg> lbm_driver: lbm_lora_send: Expected airtime: 289 ms
....
[00:10:01.627,563] <inf> lora_send: Data sent 5!
[00:10:02.627,624] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0
[00:10:02.627,655] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
[00:10:02.628,936] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
[00:10:02.629,058] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
[00:10:02.629,180] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8f DATA_LEN=0
[00:10:02.629,302] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0e DATA_LEN=12
[00:10:02.629,516] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x83 DATA_LEN=0
[00:10:02.629,608] <dbg> lbm_driver: lbm_lora_send: Expected airtime: 289 ms
[00:10:02.918,579] <dbg> lbm_driver: sx126x_dio1_callback: 
[00:10:02.918,640] <dbg> lbm_driver: op_done_work_handler: op_done_work_handler: 1
[00:10:02.918,670] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x12 DATA_LEN=2
[00:10:02.918,792] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x02 DATA_LEN=0
[00:10:02.918,884] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
[00:10:02.918,975] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP
[00:10:02.920,104] <inf> lora_send: Data sent 6!
[00:10:03.920,166] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0
[00:10:03.920,196] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
[00:10:03.921,478] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
[00:10:03.921,600] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
[00:10:03.921,722] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8f DATA_LEN=0
[00:10:03.921,844] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0e DATA_LEN=12
[00:10:03.922,058] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x83 DATA_LEN=0
[00:10:03.922,149] <dbg> lbm_driver: lbm_lora_send: Expected airtime: 289 ms
[00:10:04.211,120] <dbg> lbm_driver: sx126x_dio1_callback: 
[00:10:04.211,181] <dbg> lbm_driver: op_done_work_handler: op_done_work_handler: 1
[00:10:04.211,212] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x12 DATA_LEN=2
[00:10:04.211,334] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x02 DATA_LEN=0
[00:10:04.211,456] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
[00:10:04.211,547] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP
[00:10:04.212,646] <inf> lora_send: Data sent 7!
[00:10:05.212,707] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0
[00:10:05.212,738] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
[00:10:05.214,019] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
[00:10:05.214,141] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
[00:10:05.214,263] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8f DATA_LEN=0
[00:10:05.214,385] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0e DATA_LEN=12
[00:10:05.214,599] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x83 DATA_LEN=0
[00:10:05.214,691] <dbg> lbm_driver: lbm_lora_send: Expected airtime: 289 ms
[00:10:05.503,662] <dbg> lbm_driver: sx126x_dio1_callback: 
[00:10:05.503,723] <dbg> lbm_driver: op_done_work_handler: op_done_work_handler: 1
[00:10:05.503,753] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x12 DATA_LEN=2
[00:10:05.503,875] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x02 DATA_LEN=0
[00:10:05.503,997] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
```

Tests have been made on a custom board featuring sx1262.

```
	lora: lora@0 {
		compatible = "semtech,sx1262";
		reg = <0>;
		reset-gpios = <&gpio0 19 GPIO_ACTIVE_LOW>;
		busy-gpios = <&gpio0 20 GPIO_ACTIVE_HIGH>;
		dio1-gpios = <&gpio0 21 GPIO_ACTIVE_HIGH>;
		antenna-enable-gpios = <&gpio0 23 GPIO_ACTIVE_HIGH>;
		dio2-tx-enable;
		tcxo-power-startup-delay-ms = <5>;
		spi-max-frequency = <1000000>;
	};
```

I just forgot to bring another extra board to completely test the tx/rx modes.

sophiekovalevsky commented:
Packets are arriving on receiver:

```
*** Booting Zephyr OS build v4.1.0-3301-g4d204c5de3c0 ***
[00:00:00.354,827] <inf> lora_receive: Synchronous reception
[00:01:10.010,498] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 8 dB
[00:01:10.010,528] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 30             |hellowor ld 0    
[00:01:11.303,070] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 9 dB
[00:01:11.303,100] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 31             |hellowor ld 1    
[00:01:12.595,642] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 9 dB
[00:01:12.595,672] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 32             |hellowor ld 2    
[00:01:13.888,153] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 9 dB
[00:01:13.888,183] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 33             |hellowor ld 3    
[00:01:13.888,183] <inf> lora_receive: Asynchronous reception
[00:01:15.179,931] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 8 dB
[00:01:15.179,962] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 34             |hellowor ld 4    
[00:01:16.472,503] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 9 dB
[00:01:16.472,534] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 35             |hellowor ld 5    
[00:01:17.765,014] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 9 dB
[00:01:17.765,045] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 36             |hellowor ld 6    
[00:01:19.057,617] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 10 dB
[00:01:19.057,647] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 37             |hellowor ld 7    
[00:01:20.350,128] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 9 dB
[00:01:20.350,158] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 38             |hellowor ld 8    
[00:01:21.642,669] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 9 dB
[00:01:21.642,700] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 39             |hellowor ld 9    
[00:01:22.935,211] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 8 dB
[00:01:22.935,241] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 30             |hellowor ld 0    
[00:01:24.227,722] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 9 dB
[00:01:24.227,752] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 31             |hellowor ld 1    
[00:01:25.520,263] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 9 dB
[00:01:25.520,294] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 32             |hellowor ld 2    
[00:01:26.812,774] <inf> lora_receive: LoRa RX RSSI: -52 dBm, SNR: 8 dB
[00:01:26.812,805] <inf> lora_receive: LoRa RX payload
                                       68 65 6c 6c 6f 77 6f 72  6c 64 20 33             |hellowor ld 3    
[00:01:26.812,805] <inf> lora_receive: Stopping packet receptions
```

JordanYates commented:
PR updated with official Zephyr fork: https://github.com/zephyrproject-rtos/lora-basics-modem

mirzasadiqassb commented:
> So I've just tested this PR on a board with a TCXO and a SX1262 definition that looks almost identical to yours [Seeed-WIO](https://www.seeedstudio.com/Wio-SX1262-Wireless-Module-p-5981.html) and transmissions work fine.
> 
> ```
> 	wio_sx1262: sx1262@2 {
> 		compatible = "semtech,sx1262";
> 		status = "okay";
> 		reg = <2>;
> 		spi-max-frequency = <4000000>;
> 		reset-gpios = <&gpio1 8 GPIO_ACTIVE_LOW>;
> 		busy-gpios = <&gpio1 14 GPIO_ACTIVE_HIGH>;
> 		antenna-enable-gpios = <&gpio0 4 GPIO_ACTIVE_HIGH>;
> 		dio1-gpios = <&gpio1 13 GPIO_ACTIVE_HIGH>;
> 		dio2-tx-enable;
> 		dio3-tcxo-voltage = <SX126X_DIO3_TCXO_1V8>;
> 		tcxo-power-startup-delay-ms = <5>;
> 	};
> ```

Huh, that's weird. For your reference I am using this board [Heltec Wi-Fi Lora32 (V3)](https://heltec.org/project/wifi-lora-32-v3/). When I dig further I found out that my board got stuck during the `sx126x_set_lora_pkt_params()`. 

Here is the log:

```
I (142) spi_flash: flash io: dio
W (142) spi_flash: Detected size(8192k) larger than the size in the binary image header(4096k). Using the size in the binary image header.
--->sx126x_reset called
[00:00:00.187,000] <dbg> lbm_driver: sx126x_hal_reset: 
--->sx126x_init_retention_list called
--->sx126x_read_register called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=9
--->sx126x_write_register called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=9
--->sx126x_set_reg_mode called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x96 DATA_LEN=0
--->sx126x_set_dio2_as_rf_sw_ctrl called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x9d DATA_LEN=0
--->sx126x_convert_timeout_in_ms_to_rtc_step called
--->sx126x_set_dio3_as_tcxo_ctrl called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x97 DATA_LEN=0
--->sx126x_cal called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x89 DATA_LEN=0
--->sx126x_cfg_rx_boosted called
--->sx126x_write_register called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
--->sx126x_set_dio_irq_params called
[00:00:00.218,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x08 DATA_LEN=0
--->sx126x_set_standby called
[00:00:00.218,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x80 DATA_LEN=0
--->sx126x_set_sleep called
[00:00:00.218,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
[00:00:00.218,000] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP
*** Booting Zephyr OS build v4.1.0-2655-g2fd0ad6b8b6b ***
--->sx126x_stop_timer_on_preamble called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x9f DATA_LEN=0
[00:00:00.219,000] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
--->sx126x_set_lora_symb_nb_timeout called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0xa0 DATA_LEN=0
--->sx126x_set_pkt_type called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8a DATA_LEN=0
--->sx126x_set_rf_freq called
--->sx126x_convert_freq_in_hz_to_pll_step called
--->sx126x_set_rf_freq_in_pll_steps called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x86 DATA_LEN=0
--->sx126x_cfg_tx_clamp called
--->sx126x_read_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
--->sx126x_write_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
--->sx126x_set_pa_cfg called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x95 DATA_LEN=0
--->sx126x_set_tx_params called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8e DATA_LEN=0
--->sx126x_set_lora_mod_params called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8b DATA_LEN=0
--->sx126x_tx_modulation_workaround called
--->sx126x_read_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
--->sx126x_write_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
--->sx126x_set_lora_pkt_params called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0
--->sx126x_read_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
--->sx126x_write_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
--->sx126x_set_lora_sync_word called
--->sx126x_read_register called
[00:00:00.220,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=2
--->sx126x_write_register called
[00:00:00.221,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=2
--->sx126x_set_sleep called
[00:00:00.221,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
[00:00:00.221,000] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP
--->sx126x_set_lora_pkt_params called
[00:00:00.222,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0
[00:00:00.222,000] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
[00:00:05.222,000] <err> lbm_driver: DEVICE NOT READY FOR WRITE
--->sx126x_set_sleep called
[00:00:05.222,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
[00:00:10.223,000] <err> lbm_driver: DEVICE NOT READY FOR WRITE
[00:00:10.223,000] <err> lbm_driver: lbm_lora_send_async fail status: -22
[00:00:10.223,000] <err> lora_send: LoRa send failed
```

After further investigation, I found that adding `ral_set_standby()` in `lbm_lora_config()`, just before releasing the modem, allowed me to get past the previous issue. However, DIO1 is still never triggered, which causes the LoRa send to fail.

Here’s the log:
```
I (142) spi_flash: flash io: dio
W (142) spi_flash: Detected size(8192k) larger than the size in the binary image header(4096k). Using the size in the binary image header.
--->sx126x_reset called
[00:00:00.187,000] <dbg> lbm_driver: sx126x_hal_reset: 
--->sx126x_init_retention_list called
--->sx126x_read_register called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=9
--->sx126x_write_register called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=9
--->sx126x_set_reg_mode called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x96 DATA_LEN=0
--->sx126x_set_dio2_as_rf_sw_ctrl called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x9d DATA_LEN=0
--->sx126x_convert_timeout_in_ms_to_rtc_step called
--->sx126x_set_dio3_as_tcxo_ctrl called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x97 DATA_LEN=0
--->sx126x_cal called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x89 DATA_LEN=0
--->sx126x_cfg_rx_boosted called
--->sx126x_write_register called
[00:00:00.217,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
--->sx126x_set_dio_irq_params called
[00:00:00.218,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x08 DATA_LEN=0
--->sx126x_set_standby called
[00:00:00.218,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x80 DATA_LEN=0
--->sx126x_set_sleep called
[00:00:00.218,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
[00:00:00.218,000] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP
*** Booting Zephyr OS build v4.1.0-2655-g2fd0ad6b8b6b ***
--->sx126x_stop_timer_on_preamble called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x9f DATA_LEN=0
[00:00:00.219,000] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
--->sx126x_set_lora_symb_nb_timeout called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0xa0 DATA_LEN=0
--->sx126x_set_pkt_type called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8a DATA_LEN=0
--->sx126x_set_rf_freq called
--->sx126x_convert_freq_in_hz_to_pll_step called
--->sx126x_set_rf_freq_in_pll_steps called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x86 DATA_LEN=0
--->sx126x_cfg_tx_clamp called
--->sx126x_read_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
--->sx126x_write_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
--->sx126x_set_pa_cfg called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x95 DATA_LEN=0
--->sx126x_set_tx_params called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8e DATA_LEN=0
--->sx126x_set_lora_mod_params called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8b DATA_LEN=0
--->sx126x_tx_modulation_workaround called
--->sx126x_read_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
--->sx126x_write_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
--->sx126x_set_lora_pkt_params called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0
--->sx126x_read_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
--->sx126x_write_register called
[00:00:00.219,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
--->sx126x_set_lora_sync_word called
--->sx126x_read_register called
[00:00:00.220,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=2
--->sx126x_write_register called
[00:00:00.221,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=2
--->sx126x_set_standby called
[00:00:00.221,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x80 DATA_LEN=0
--->sx126x_set_sleep called
[00:00:00.221,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
[00:00:00.221,000] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP
--->sx126x_set_lora_pkt_params called
[00:00:00.222,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8c DATA_LEN=0
[00:00:00.222,000] <dbg> lbm_driver: sx126x_ensure_device_ready: SLEEP -> ACTIVE
--->sx126x_read_register called
[00:00:00.223,000] <dbg> lbm_driver: sx126x_hal_read: CMD[0]=0x1d DATA_LEN=1
--->sx126x_write_register called
[00:00:00.223,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0d DATA_LEN=1
--->sx126x_set_buffer_base_address called
[00:00:00.224,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x8f DATA_LEN=0
--->sx126x_write_buffer called
[00:00:00.224,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x0e DATA_LEN=10
--->sx126x_convert_timeout_in_ms_to_rtc_step called
--->sx126x_set_tx called
--->sx126x_convert_timeout_in_ms_to_rtc_step called
--->sx126x_set_tx_with_timeout_in_rtc_step called
[00:00:00.224,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x83 DATA_LEN=0
[00:00:00.224,000] <dbg> lbm_driver: sx126x_hal_write: TX SET
[00:00:00.224,000] <dbg> lbm_driver: sx126x_hal_write: DIO1 level: 0
--->sx126x_get_lora_time_on_air_in_ms called
--->sx126x_get_lora_time_on_air_numerator called
--->sx126x_get_lora_bw_in_hz called
[00:00:00.224,000] <dbg> lbm_driver: lbm_lora_send: Expected airtime: 289 ms
--->sx126x_set_sleep called
[00:00:00.802,000] <dbg> lbm_driver: sx126x_hal_write: CMD[0]=0x84 DATA_LEN=0
[00:00:00.802,000] <dbg> lbm_driver: sx126x_hal_write: ACTIVE -> SLEEP
[00:00:00.803,000] <err> lbm_driver: Packet transmission failed!
[00:00:00.803,000] <err> lora_send: LoRa send failed
```

Also, I’m not sure if this could be related, but could you share your SPI devicetree node just in case?

kartben commented:
@JordanYates should this be labeled as "4.2"?

JordanYates commented:
> @JordanYates should this be labeled as "4.2"?

That depends on whether we can get enough reviews and resolve the observed issues.
I would guess that the feature freeze is too soon to make it in.

kartben commented:
OK, fair, and thanks!

JordanYates committed:
lora: lbm: enable `SX126x` RX boost option

Add the option to enable the RX boost mode in devicetree.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
lora: lbm: choice of reported RSSI type

Add a choice of which RSSI type should be reported to the application
layer, packet RSSI (which stops at the noise floor) or signal RSSI
(which estimates the LoRa signal strength into the noise floor).

Signed-off-by: Jordan Yates <jordan@embeint.com>

thedjnK commented on drivers/lora/lora_basics_modem/Kconfig:
```diff
@@ -0,0 +1,37 @@
+	int "Maximum size payload that can be received by async RX"
+	default 256
+	help
+		A buffer of this size is allocated on the system workqueue stack when
+		running the async RX handler.
```
1x tab and 2x spaces

thedjnK commented on drivers/lora/lora_basics_modem/lbm_sx126x.c:
```diff
@@ -0,0 +1,466 @@
+	const struct device *dev = context;
+	const struct lbm_sx126x_config *config = dev->config;
+
+	LOG_DBG("");
```
?

thedjnK commented on drivers/lora/lora_basics_modem/lbm_sx126x.c:
```diff
@@ -0,0 +1,466 @@
+	const struct device *dev = context;
+	int ret;
+
+	LOG_DBG("");
```
seems to be a few of these

thedjnK commented on drivers/lora/loramac_node/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+# library. Hence, if it exists then the source files are added to that otherwise
+# a library with same name is created.
+if(TARGET loramac-node)
+	set(ZEPHYR_CURRENT_LIBRARY loramac-node)
```
2 space indent for cmake

thedjnK commented on drivers/lora/loramac_node/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+zephyr_library_sources_ifdef(CONFIG_HAS_SEMTECH_RADIO_DRIVERS sx12xx_common.c)
+zephyr_library_sources_ifdef(CONFIG_LORA_SX127X sx127x.c)
+
+if (CONFIG_LORA_SX126X OR CONFIG_LORA_STM32WL_SUBGHZ_RADIO)
```
no space between if and bracket

thedjnK commented on modules/lora-basics-modem/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+
+  if(CONFIG_LORA_SX126X)
+    include(sx126x.cmake)
+  endif()
+  if(CONFIG_LORA_SX127X)
```
nit. `elseif()`

thedjnK reviewed (commented):
very quick skim through

rijesha commented on drivers/lora/lora_basics_modem/lbm_common.c:
```diff
@@ -0,0 +1,552 @@
+	 * a failed transmission, and not some minor timing variation between
+	 * modem and driver.
+	 */
+	ret = k_poll(&evt, 1, K_MSEC(2 * air_time));
```
I have been testing out this new driver. 
In this line here, I seem to get send failures close to 50% of the time often even though I successfully receive the message on the other end. Line 210 gets triggered as a result and the return from the send function is -11.
If I change  the wait time to 3 * air_time it works well without send errors.

My lora settings are: 
```
    config.frequency = 915100000;
    config.bandwidth = BW_500_KHZ;
    config.datarate = SF_6;
    config.preamble_len = 8;
    config.coding_rate = CR_4_5;
    config.iq_inverted = false;
    config.public_network = false;
    config.tx_power = 14;
```

rijesha replied:
msg_len = 9 air_time = 6.

JordanYates replied:
Pushed an update with a 10ms offset. I've left it as 2x since 1.5x requires either a division or float multiplication, neither of which are fast. And this is only the failure path, under normal operation it won't timeout anyway.

fabiobaltieri commented on drivers/lora/lora_basics_modem/lbm_common.c:
```diff
@@ -0,0 +1,549 @@
+	ret = k_poll(&evt, 1, K_MSEC(10 + (2 * air_time)));
+	if (ret < 0) {
+		if (modem_release(dev)) {
+			LOG_ERR("Packet transmission failed!");
```
log ret?

JordanYates replied:
I'm pretty sure the only error code from `k_poll` is `-EAGAIN`, which doesn't provide any useful information to the user.

fabiobaltieri commented on drivers/lora/lora_basics_modem/lbm_common.c:
```diff
@@ -0,0 +1,549 @@
+	ret = k_poll(&evt, 1, timeout);
+	if (ret < 0) {
+		if (modem_release(dev)) {
+			LOG_INF("Receive timeout");
```
log ret, should this be a LOG_WRN or LOG_ERR?

JordanYates replied:
Again, the timeout is the only failing return code, which is adequately described by the message itself.

I don't think this needs to be `WRN` or `ERR`, since there is no guarantee that the caller actually expected to receive a message (it could just be a trailing RX window after TX). If the user considers this an error, they can add their own logging at the application layer.

fabiobaltieri commented on drivers/lora/lora_basics_modem/lbm_common.c:
```diff
@@ -0,0 +1,549 @@
+
+	/* Retrieve the packet payload */
+	status = ral_get_pkt_payload(&config->ralf.ral, sizeof(rx_buffer), rx_buffer, &size);
+	if (status == RAL_STATUS_OK) {
+		LOG_HEXDUMP_DBG(rx_buffer, size, "RX");
+	} else {
```
maybe swap the condition and put the `LOG_HEXDUMP_DBG` unindented, this branch with just the dbg looks weird

fabiobaltieri commented on drivers/lora/lora_basics_modem/lbm_common.c:
```diff
@@ -0,0 +1,549 @@
+	bool error_irq;
+	int ret = 0;
+
+	LOG_DBG("%s: %d", __func__, data->modem_mode);
```
doesn't LOG_DBG log __func__ automatically?

fabiobaltieri commented on drivers/lora/lora_basics_modem/lbm_common.c:
```diff
@@ -0,0 +1,549 @@
+
+	switch (data->modem_mode) {
+	case MODE_SLEEP:
+		LOG_WRN("Unexpected modem mode");
```
log data->modem_mode

fabiobaltieri commented on drivers/lora/lora_basics_modem/lbm_common.c:
```diff
@@ -0,0 +1,549 @@
+
+	/* Enable all relevant interrupts */
+	status = ral_set_dio_irq_params(&config->ralf.ral,
+					RAL_IRQ_TX_DONE | RAL_IRQ_RX_DONE | RAL_IRQ_RX_HDR_ERROR |
+						RAL_IRQ_RX_CRC_ERROR | RAL_IRQ_CAD_DONE |
+						RAL_IRQ_CAD_OK);
```
indentation looks off, you may want to consider having these in a define as well maybe

fabiobaltieri commented on drivers/lora/lora_basics_modem/lbm_sx126x.c:
```diff
@@ -0,0 +1,462 @@
+{
+	const struct lbm_sx126x_config *config = dev->config;
+	struct lbm_sx126x_data *data = dev->data;
+	uint8_t get_status_cmd[2] = {0xC0, 0xFF};
```
is there a define for that 0xc0?

fabiobaltieri commented on drivers/lora/lora_basics_modem/lbm_sx126x.c:
```diff
@@ -0,0 +1,462 @@
+	}
+
+	/* 0x84 - SX126x_SET_SLEEP opcode */
+	if (command[0] == 0x84) {
```
define for 0x84?

fabiobaltieri commented on drivers/lora/lora_basics_modem/lbm_sx126x.c:
```diff
@@ -0,0 +1,462 @@
+	output_params->pa_cfg.pa_lut = 0x01;
+
+	if (config->variant == VARIANT_SX1261) {
+		power = MAX(power, -17);
+		power = MIN(power, 15);
```
add  defines for min/max for the various chips, also you can use `CLAMP` here
